peer_test.go

```diff
@@ -217,3 +219,268 @@ func TestPeerSelectionPreferIncoming(t *testing.T) {
+	client   *Channel
+}
+
+// setupServers will create numPeer servers, and register handlers on them.
+func (pt *peerSelectionTest) setupServers(t *testing.T) {
+	pt.servers = make([]*Channel, pt.numPeers)
+
+	// Set up numPeers servers.
+	for i := 0; i < pt.numPeers; i++ {
+		pt.servers[i], _ = pt.NewService(t, "hyperbahn")
+		pt.servers[i].Register(raw.Wrap(newTestHandler(pt.t)), "echo")
+	}
+	fmt.Println("all server nodes setup")
+}
+
+func (pt *peerSelectionTest) setupAffinity(t *testing.T) {
+	var wg sync.WaitGroup
+	pt.affinity = make([]*Channel, pt.numAffinity)
+	for i := range pt.affinity {
+		pt.affinity[i] = pt.servers[i]
+	}
+
+	for _, server := range pt.servers {
+		pt.client.Peers().Add(server.PeerInfo().HostPort)
+	}
+	fmt.Println("added all hyperbahn nodes")
+
+	wg.Add(pt.numAffinity)
+	// Connect from the affinity nodes to the service.
+	for _, affinity := range pt.affinity {
+		affinity.Peers().Add(pt.client.PeerInfo().HostPort)
+		go func(affinity *Channel) {
+			pt.makeCall(affinity.GetSubChannel(pt.client.PeerInfo().ServiceName))
+			wg.Done()
+		}(affinity)
+	}
+	wg.Wait()
+	fmt.Println("affinity nodes connected")
+}
+
+func (pt *peerSelectionTest) setupClient(t *testing.T) {
+	pt.client, _ = pt.NewService(t, "client")
+	pt.client.Register(raw.Wrap(newTestHandler(t)), "echo")
+	for _, server := range pt.servers {
+		pt.client.Peers().Add(server.PeerInfo().HostPort)
+	}
+	time.Sleep(time.Millisecond * 100)
+}
+
+func (pt *peerSelectionTest) makeCall(sc *SubChannel) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+	_, _, _, err := raw.CallSC(ctx, sc, "echo", nil, nil)
+	require.NoError(pt.t, err, "raw.Call failed")
+}
+
+func (pt *peerSelectionTest) runStress() {
+	var wg sync.WaitGroup
+
+	// helper that will make a request every n ticks.
+	reqEveryNTicks := func(n int, sc *SubChannel, clock <-chan struct{}) {
+		defer wg.Done()
+
+		for {
+			for i := 0; i < n; i++ {
+				_, ok := <-clock
+				if !ok {
+					return
+				}
+			}
+			pt.makeCall(sc)
+		}
+	}
+
+	numClock := pt.numConcurrent + pt.numAffinity
+	clocks := make([]chan struct{}, numClock)
+	for i := 0; i < numClock; i++ {
+		clocks[i] = make(chan struct{})
+	}
+	wg.Add(numClock)
+
+	// server outbound request
+	sc := pt.client.GetSubChannel("hyperbahn")
+	for i := 0; i < pt.numConcurrent; i++ {
+		go reqEveryNTicks(1, sc, clocks[i])
+	}
+	// affinity incoming requests
+	for i := 0; i < pt.numAffinity; i++ {
+		go reqEveryNTicks(1, pt.affinity[i].GetSubChannel(pt.client.PeerInfo().ServiceName), clocks[i+pt.numConcurrent])
+	}
+
+	tickAllClocks := func() {
+		for i := 0; i < numClock; i++ {
+			clocks[i] <- struct{}{}
+		}
+	}
+
+	const numTicks = 10000
+	for i := 0; i < numTicks; i++ {
+		if i%(numTicks/10) == 0 {
+			fmt.Println("upto", i, (100.0 * float64(i) / numTicks))
+		}
+		tickAllClocks()
+	}
+
+	time.Sleep(time.Millisecond * 100)
+	for i := 0; i < numClock; i++ {
+		close(clocks[i])
+	}
+	wg.Wait()
+}
+
+func PeersHeapPerf(t *testing.T) {
```

Comment:
since this doesn't have a `Test` prefix, it won't even be run.

I would suggest doing something like this:

```go
func TestPeerHeapBalance(t *testing.T) {
  tests := []struct {
    numHyperbahn int
    affinityRatio float64
    numConcurrent int
    hasInboundCall bool
  }{
    {
      numHyperbahn: 1000,
      affinityRatio: 0.1,
      numConcurrent: 5,
      hasInboundCall: true,
    },
    {
      numHyperbahn: 100,
      affinityRatio: 0.1,
      numConcurrent: 1,
      hasInboundCall: true,
    },
    {
      numHyperbahn: 100,
      affinityRatio: 0,
      numConcurrent: 1,
      hasInboundCall: true,
    },
  }

  for _, tt := range tests {
    peerHeapStressTest(t, tt.numHyperbahn, tt.affinityRatio, tt.numConcurrent, tt.hasInboundCall)
  }
}
```

(The actual test cases will need tweaking, but it should be easy to add new cases to run, and we should have a few there)